File: ios/GradeThread/Persistence/PersistenceHealth.swift
import Foundation

/// Makes local SwiftData write failures observable instead of silently
/// swallowed (US-1142).
///
/// The offline cache used to scatter `try? context.save()` across the sync,
/// upload, and editing surfaces — a failed save (disk full, store corruption,
/// truncated write under pressure) vanished with no breadcrumb and no user
/// signal, so the UI showed edits that were never persisted. ``ModelContext``'s
/// `saveOrLog(_:)` now routes every failure here, which:
///
///   1. records a scrubbed Sentry breadcrumb (always — diagnostics), and
///   2. posts a main-thread notification the app shell throttles into a
///      one-time "couldn't save on this device" notice once writes are
///      *persistently* failing (not on a single transient blip).
enum PersistenceHealth {

    /// Posted whenever a local save fails. The shell observes this (debounced to
    /// the main run loop) and surfaces a throttled notice; tests assert it fires.
    /// Posting thread is whatever called `saveOrLog`; observers must hop to main.
    static let saveFailedNotification = Notification.Name("com.gradethread.app.localSaveFailed")

    /// Records a local-save failure: a scrubbed breadcrumb plus a notification.
    /// `nonisolated` so the `SyncMergeActor`, the nonisolated
    /// `OfflineMutationQueue`, and `@MainActor` stores can all call it from
    /// within their own isolation with no hop at the call site.
    nonisolated static func recordSaveFailure(operation: String, error: Error) {
        Telemetry.backgroundBreadcrumb(
            "SwiftData save failed [\(operation)]: \(error.localizedDescription)",
            category: "persistence"
        )
        #if DEBUG
        print("[Persistence] save failed [\(operation)]: \(error)")
        #endif
        NotificationCenter.default.post(
            name: saveFailedNotification,
            object: nil,
            userInfo: ["operation": operation]
        )
    }
}

File: android/app/src/main/java/com/gradethread/app/sync/PersistenceHealth.kt
package com.gradethread.app.sync

import com.gradethread.app.platform.telemetry.Telemetry
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import java.util.concurrent.atomic.AtomicInteger

/**
 * US-1322 / US-1142 (iOS PersistenceHealth): a Room write failing (disk
 * full, corruption under pressure) must never vanish silently. Every failure
 * records a SCRUBBED breadcrumb (always — diagnostics); once failures are
 * PERSISTENT (>= [NOTICE_THRESHOLD], not a single transient blip) a one-time
 * "couldn't save on this device" notice surfaces. Acknowledging resets the
 * counter so the notice can re-arm if trouble continues.
 */
object PersistenceHealth {

    const val NOTICE_THRESHOLD = 3

    private val failureCount = AtomicInteger(0)
    private val noticeFlow = MutableStateFlow(false)

    /** The shell binds an alert to this ("couldn't save on this device"). */
    val noticeNeeded: StateFlow<Boolean> = noticeFlow

    /** Record a local-save failure from any thread. */
    fun recordSaveFailure(operation: String, error: Throwable) {
        Telemetry.breadcrumb(
            "Room save failed [$operation]: ${error.message ?: error.javaClass.simpleName}",
            category = "persistence",
        )
        if (failureCount.incrementAndGet() >= NOTICE_THRESHOLD) {
            noticeFlow.value = true
        }
    }

    /** User acknowledged — reset so the notice re-arms only on NEW trouble. */
    fun acknowledgeNotice() {
        failureCount.set(0)
        noticeFlow.value = false
    }

    /** Test seam. */
    internal fun resetForTest() = acknowledgeNotice()
}
